# **A. Modelo de Regresión con Keras**

### **1. Tema de la asignación**

En este proyecto, construiremos un modelo de **regresión** utilizando la biblioteca **Keras** para modelar los datos sobre la **resistencia a la compresión del hormigón**. Este ejercicio es una continuación del análisis realizado en el **Laboratorio 3**, aplicando redes neuronales artificiales para predecir la resistencia del material en función de sus componentes.

---

### **2. Datos del hormigón**

Para su comodidad, los datos se pueden encontrar en el siguiente enlace:

🔗 [Descargar datos](https://cocl.us/concrete_data)

Los datos contienen información sobre la resistencia del hormigón en función de los siguientes predictores:

- **Cemento**
- **Escoria de alto horno**
- **Cenizas volantes**
- **Agua**
- **Superplastificante**
- **Árido grueso**
- **Árido fino**

El objetivo es construir un modelo que prediga la **resistencia a la compresión** del hormigón basándose en estos factores.

---

### **3. Instrucciones para la tarea**

1. **Carga de datos**: Importar y visualizar el conjunto de datos para entender su estructura.
2. **Preprocesamiento**: Normalización y división en conjuntos de entrenamiento y prueba.
3. **Construcción del modelo**: Implementación de una red neuronal con Keras para regresión.
4. **Entrenamiento y evaluación**: Comparar el rendimiento del modelo a través de métricas adecuadas, como el error medio cuadrático (*MSE*).
5. **Discusión de resultados**: Comparar la diferencia en la media de los errores obtenidos entre distintas configuraciones del modelo.

---


### **4. Código de implementación**

In [1]:
# Importar bibliotecas necesarias

# pandas: Librería para manipulación y análisis de datos estructurados (DataFrames y Series)
import pandas as pd

# numpy: Librería para operaciones numéricas y manejo de matrices/arreglos
import numpy as np

# tensorflow: Marco de trabajo para el desarrollo de modelos de aprendizaje profundo
import tensorflow as tf

# keras: API de alto nivel dentro de TensorFlow para construir y entrenar redes neuronales
from tensorflow import keras

# Sequential: Clase de Keras que permite construir modelos de redes neuronales capa por capa
from tensorflow.keras.models import Sequential

# Dense: Capa de neuronas totalmente conectada utilizada en redes neuronales artificiales
from tensorflow.keras.layers import Dense

# train_test_split: Función de Scikit-learn para dividir los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

# StandardScaler: Método para normalizar los datos asegurando que tengan media 0 y desviación estándar 1
from sklearn.preprocessing import StandardScaler

# mean_squared_error: Métrica para evaluar la precisión del modelo comparando valores predichos con los reales
from sklearn.metrics import mean_squared_error

In [7]:
# Importar la función Input de Keras para definir explícitamente la capa de entrada
from tensorflow.keras import Input

In [2]:
# Cargar los datos
data_url = "https://cocl.us/concrete_data"
df = pd.read_csv(data_url)

In [3]:
# Separar variables predictoras y variable objetivo
X = df.drop(columns=['Strength'])  # Variables de entrada
y = df['Strength']  # Variable objetivo

In [12]:
# Inicializar lista para almacenar los errores
mse_list = []

In [14]:
# Repetir el proceso 50 veces
for _ in range(50):
    # Dividir los datos en conjunto de entrenamiento (70%) y prueba (30%)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Normalizar los datos
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # Construcción del modelo con la recomendación de Keras
    model = Sequential([
        Input(shape=(X_train.shape[1],)),  # Definir la capa de entrada explícitamente
        Dense(64, activation='relu'),  # Primera capa oculta con 64 neuronas y ReLU
        Dense(64, activation='relu'),  # Segunda capa oculta con 64 neuronas y ReLU
        Dense(1)  # Capa de salida con 1 neurona para regresión
    ])

    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse')

    # Entrenar el modelo con 50 épocas
    model.fit(X_train, y_train, epochs=50, verbose=0, batch_size=10)

    # Evaluación del modelo
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10

In [15]:
# Calcular la media y desviación estándar de los errores
mse_mean = np.mean(mse_list)
mse_std = np.std(mse_list)

In [16]:
print(f'Error Medio Cuadrático promedio: {mse_mean:.4f}')
print(f'Desviación estándar del MSE: {mse_std:.4f}')

Error Medio Cuadrático promedio: 49.6710
Desviación estándar del MSE: 25.4673


## Conclusión

En este experimento, se construyó un modelo de red neuronal para predecir la resistencia a la compresión del hormigón a partir de sus componentes. Tras realizar 50 repeticiones del proceso de entrenamiento y evaluación, se obtuvo un error medio cuadrático (MSE) promedio de **49.6710** con una desviación estándar de **25.4673**.

Estos resultados indican que el modelo tiene un desempeño estable; sin embargo, la alta desviación estándar sugiere que la precisión de las predicciones puede verse afectada por la variabilidad en la selección del conjunto de entrenamiento. Para mejorar su rendimiento, podrían implementarse estrategias como el ajuste de hiperparámetros, el uso de arquitecturas más complejas o el aumento del tamaño del conjunto de datos.

La red neuronal construida logra modelar la relación entre los componentes del hormigón y su resistencia, pero aún presenta margen de mejora para optimizar su precisión y reducir la variabilidad en sus predicciones.


#**B. Implementación con Datos Normalizados**



In [17]:
# Inicializar lista para almacenar los errores con datos normalizados
mse_list_normalized = []

In [19]:
# Repetir el proceso 50 veces con datos normalizados
for _ in range(50):
    # Dividir los datos en conjunto de entrenamiento (70%) y prueba (30%)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Normalizar los datos (restar la media y dividir por la desviación estándar)
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # Construcción del modelo corregido
    model = Sequential([
        Input(shape=(X_train.shape[1],)),  # Definir la entrada explícitamente
        Dense(10, activation='relu'),  # Primera capa oculta
        Dense(10, activation='relu'),  # Segunda capa oculta
        Dense(10, activation='relu'),  # Tercera capa oculta
        Dense(1)  # Capa de salida para regresión
    ])

    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse')

    # Entrenar el modelo con 50 épocas
    model.fit(X_train, y_train, epochs=50, verbose=0, batch_size=10)

    # Evaluación del modelo
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list_normalized.append(mse)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━

In [20]:
# Calcular la media y desviación estándar de los errores con datos normalizados
mse_mean_normalized = np.mean(mse_list_normalized)
mse_std_normalized = np.std(mse_list_normalized)

print(f'Error Medio Cuadrático promedio con normalización: {mse_mean_normalized:.4f}')
print(f'Desviación estándar del MSE con normalización: {mse_std_normalized:.4f}')

Error Medio Cuadrático promedio con normalización: 89.6288
Desviación estándar del MSE con normalización: 22.6291


### Comparación de los resultados

Comparando los errores medios al cuadrado de la primera implementación y de la versión con datos normalizados:

### Modelo sin normalización:
- **MSE promedio:** 49.6710  
- **Desviación estándar:** 25.4673  

### Modelo con normalización:
- **MSE promedio:** 89.6288  
- **Desviación estándar:** 22.6291  

## **Conclusión**
En este caso, la normalización de los datos no mejoró el desempeño del modelo, ya que el MSE promedio aumentó de **49.6710** a **89.6288**. Aunque la desviación estándar disminuyó ligeramente, lo que indica una menor variabilidad en los errores, el incremento del MSE sugiere que la normalización no favoreció la capacidad predictiva de la red neuronal en esta implementación.  

La normalización suele mejorar la estabilidad y la convergencia del modelo en muchos escenarios, pero su efectividad depende del tipo de datos y del modelo utilizado. Para mejorar el rendimiento, podría ser necesario ajustar otros hiperparámetros, explorar diferentes estrategias de normalización o probar arquitecturas más complejas.


## **C. Implementación con 100 Épocas**

In [33]:
# Inicializar lista para almacenar los errores con 100 épocas
mse_list_100_epochs = []

# Repetir el proceso 50 veces con datos normalizados y 100 épocas
for _ in range(50):
    # Dividir los datos en conjunto de entrenamiento (70%) y prueba (30%)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Normalizar los datos (restar la media y dividir por la desviación estándar)
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # Construcción del modelo
    model = Sequential([
        Input(shape=(X_train.shape[1],)),  # Definir la entrada explícitamente
        Dense(10, activation='relu'),  # Capa oculta con 10 nodos y ReLU
        Dense(1)  # Capa de salida para regresión
    ])

    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse')

    # Entrenar el modelo con **100 épocas**
    model.fit(X_train, y_train, epochs=100, verbose=0, batch_size=10)

    # Evaluación del modelo
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list_100_epochs.append(mse)


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


In [34]:
# Calcular la media y desviación estándar de los errores con 100 épocas
mse_mean_100_epochs = np.mean(mse_list_100_epochs)
mse_std_100_epochs = np.std(mse_list_100_epochs)

In [35]:
print(f'Error Medio Cuadrático promedio con 100 épocas: {mse_mean_100_epochs:.4f}')
print(f'Desviación estándar del MSE con 100 épocas: {mse_std_100_epochs:.4f}')

Error Medio Cuadrático promedio con 100 épocas: 78.8172
Desviación estándar del MSE con 100 épocas: 16.1893


### Comparación de los resultados

Comparando los errores medios al cuadrado obtenidos en los pasos anteriores:

#### Modelo con 50 épocas (Paso B - Normalizado):
- **MSE promedio:** 89.6288  
- **Desviación estándar:** 22.6291  

#### Modelo con 100 épocas (Paso C):
- **MSE promedio:** 78.8172  
- **Desviación estándar:** 16.1893  

## **Conclusión**

Aumentar el número de épocas a **100** ha reducido el error medio cuadrático de **89.6288** a **78.8172**, lo que indica que el modelo ha seguido aprendiendo y mejorando su desempeño con más iteraciones. Además, la disminución en la desviación estándar (**de 22.6291 a 16.1893**) sugiere que las predicciones son más consistentes y menos sensibles a la selección del conjunto de entrenamiento.  

Sin embargo, aunque el modelo ha mejorado con más épocas, es importante evaluar si existe riesgo de **sobreajuste**. Si el error en los datos de prueba deja de disminuir o comienza a aumentar en futuras iteraciones, podría ser necesario aplicar estrategias como la **regularización** o el **early stopping** para evitar que el modelo memorice los datos en lugar de generalizar correctamente.  

En general, el aumento en las épocas ha sido beneficioso en este caso, pero se recomienda seguir monitoreando el desempeño para determinar el punto óptimo de entrenamiento.  

## **D. Implementación con tres capas ocultas**



In [36]:
# Inicializar lista para almacenar los errores con tres capas ocultas
mse_list_3_layers = []

In [40]:
# Repetir el proceso 50 veces con datos normalizados y 3 capas ocultas
for _ in range(50):
    # Dividir los datos en conjunto de entrenamiento (70%) y prueba (30%)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Normalizar los datos (restar la media y dividir por la desviación estándar)
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # Construcción del modelo con más capas ocultas
    model = Sequential([
        Input(shape=(X_train.shape[1],)),  # Definir la entrada explícitamente
        Dense(10, activation='relu'),  # Primera capa oculta
        Dense(10, activation='relu'),  # Segunda capa oculta
        Dense(10, activation='relu'),  # Tercera capa oculta
        Dense(1)  # Capa de salida para regresión
    ])

    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse')

    # Entrenar el modelo con **50 épocas**
    model.fit(X_train, y_train, epochs=50, verbose=0, batch_size=10)

    # Evaluación del modelo
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list_3_layers.append(mse)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━

In [41]:
# Calcular la media y desviación estándar de los errores con tres capas ocultas
mse_mean_3_layers = np.mean(mse_list_3_layers)
mse_std_3_layers = np.std(mse_list_3_layers)

In [42]:
print(f'Error Medio Cuadrático promedio con 3 capas ocultas: {mse_mean_3_layers:.4f}')
print(f'Desviación estándar del MSE con 3 capas ocultas: {mse_std_3_layers:.4f}')

Error Medio Cuadrático promedio con 3 capas ocultas: 83.1904
Desviación estándar del MSE con 3 capas ocultas: 25.1132


### **Comparación de los resultados**

Comparando los errores medios al cuadrado obtenidos en los pasos anteriores:

#### Modelo con 50 épocas (Paso B - Normalizado):
- **MSE promedio:** 89.6288  
- **Desviación estándar:** 22.6291  

#### Modelo con 100 épocas (Paso C):
- **MSE promedio:** 78.8172  
- **Desviación estándar:** 16.1893  

#### Modelo con 3 capas ocultas (Paso D):
- **MSE promedio:** 83.1904  
- **Desviación estándar:** 25.1132  

## **Conclusión**

Agregar **3 capas ocultas** ha resultado en un **MSE promedio de 83.1904**, lo que representa una leve mejora respecto al modelo con **50 épocas** (**89.6288**), pero un peor desempeño en comparación con el modelo entrenado por **100 épocas** (**78.8172**). Además, la **desviación estándar aumentó** de **16.1893** a **25.1132**, lo que indica que las predicciones son menos estables.  

Este resultado sugiere que, aunque una red más profunda puede capturar patrones más complejos, **no siempre garantiza una mejor generalización**. La **mayor variabilidad en los errores** podría ser indicativa de **sobreajuste**, donde el modelo se ajusta demasiado a los datos de entrenamiento sin mejorar significativamente su capacidad predictiva en datos nuevos.  

Para mejorar el rendimiento, podrían explorarse estrategias como:  
- **Ajuste de hiperparámetros**, optimizando el número de capas y neuronas.  
- **Regularización**, aplicando técnicas como **Dropout** o penalización **L2** para evitar sobreajuste.  
- **Mayor cantidad de datos**, para mejorar la capacidad de generalización del modelo.  

Si bien agregar más capas puede mejorar la capacidad de representación del modelo, **en este caso no ha superado el desempeño del modelo con 100 épocas**, por lo que se recomienda evaluar combinaciones óptimas de arquitectura y estrategias de regularización.  
